[Feature] Variational Bayesian last layer models as surrogate models

brunzema opened the pull request:
## Motivation

This PR adds variational Bayesian last layers (VBLLs) [1], which demonstrated very promising results in the context of BO in our last paper [2], to BoTorch. The goal is to provide a BoTorch-compatible implementation of VBLL surrogates for standard use cases (single-output models), making them accessible to the community as quickly as possible. This PR does not yet contain all the features discussed in [2] such as the continual learning.  If there is the interest to also add the continual learning, I am happy to add them down the line! 

The VBLLs can be used in standard acquisition functions such as (log)EI but are especially nice for Thompson sampling as the Thompson sample of a Bayesian last layer model is a differentiable standard feed forward neural network which is useful for (almost) global optimization of the sample for the next query location. 

## Implementation details

This PR adds the implementation to the community folders--also here, if there is a large interest in the model, I am happy to help merge them into the main part of the repo. The added files of this PR are the following

```
botorch_community
|-- acquisition
|   |-- bll_thompson_sampling.py # TS for Bayesian last layer models
|-- models
|   |-- vblls.py # BoTorch wrapper for VBLLs
|-- posteriors
|   |-- bll_posterior.py # Posterior class for Bayesian last layer models
notebooks_community
|-- vbll_thompson_sampling.ipynb # Tutorial on how use the VBLL model
test_community
|-- models
|   |-- test_vblls.py # test for the VBLL models functionality (backbone freezing for feature reuse, etc)
```

The current implementation build directly on the [VBLL repo](https://github.com/VectorInstitute/vbll), which is actively maintained and depends only on PyTorch. Using this repo allows improvements—e.g., better variational posterior initialization—to be directly beneficial for BO.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

The PR does not change any functionality of the current code base. The core functionality of the VBLLs should be covered by `test_vblls.py`. Let me know if further tests are required.

## Related PRs

This PR does not change functionality and I did not see any PRs regarding last layer models in BoTorch. Maybe this implementation can useful also for other BLLs.

## References

[1] P. Brunzema, M. Jordahn, J. Willes, S. Trimpe, J. Snoek, J. Harrison. [Bayesian Optimization via Continual Variational Last Layer Training](https://arxiv.org/abs/2412.09477). International Conference on Learning Representations (ICLR), 2025.

[2] J. Harrison, J. Willes, J. Snoek. [Variational Bayesian Last Layers](https://arxiv.org/abs/2404.11599). International Conference on Learning Representations (ICLR), 2024.

eytan commented:
Thanks for this contribution, Paul! This was a very intriguing paper and
it’s great to see it in botch.
We typically don’t take on new external dependencies but if it’s pure
PyTorch that should be OK,, as long as it’s possible for your package to
have some sort of unit or integration test that ensures any future changes
to your package do not cause problems for the botch implementation/
functionality demonstrated in the nb.

On Fri, Feb 21, 2025 at 9:57 AM Paul Brunzema ***@***.***>
wrote:

> Motivation
>
> This PR adds variational Bayesian last layers (VBLLs) [1], which
> demonstrated pre very promising results in the context of BO in our last
> paper [2], to BoTorch. The goal is to provide a BoTorch-compatible
> implementation of VBLL surrogates for standard use cases (single-output
> models), making them accessible to the community as quickly as possible.
> This PR does not yet contain all the features discussed in [2] such as the
> continual learning. If there is the interest to also add the continual
> learning, I am happy to add them down the line!
>
> The VBLLs can be used in standard acquisition functions such as (log)EI
> but are especially nice for Thompson sampling as the Thompson sample of a
> Bayesian last layer model is a differentiable standard feed forward neural
> network which is useful for (almost) global optimization of the sample for
> the next query location.
> Implementation details
>
> This PR adds the implementation to the community folders--also here, if
> there is a large interest in the model, I am happy to help merge them into
> the main part of the repo. The added files of this PR are the following
>
> botorch_community
> |-- acquisition
> |   |-- bll_thompson_sampling.py # TS for Bayesian last layer models
> |-- models
> |   |-- vblls.py # BoTorch wrapper for VBLLs
> |-- posteriors
> |   |-- bll_posterior.py # Posterior class for Bayesian last layer models
> notebooks_community
> |-- vbll_thompson_sampling.ipynb # Tutorial on how use the VBLL model
> test_community
> |-- models
> |   |-- test_vblls.py # test for the VBLL models functionality (backbone freezing for feature reuse, etc)
>
> The current implementation build directly on the VBLL repo
> <https://github.com/VectorInstitute/vbll>, which is actively maintained
> and depends only on PyTorch. Using this repo allows improvements—e.g.,
> better variational posterior initialization—to be directly beneficial for
> BO.
> Have you read the Contributing Guidelines on pull requests
> <https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests>
> ?
>
> Yes.
> Test Plan
>
> The PR does not change any functionality of the current code base. The
> core functionality of the VBLLs should be covered by test_vblls.py. Let
> me know if further tests are required.
> Related PRs
>
> This PR does not change functionality and I did not see any PRs regarding
> last layer models in BoTorch. Maybe this implementation can useful also for
> other BLLs.
> References
>
> [1] P. Brunzema, M. Jordahn, J. Willes, S. Trimpe, J. Snoek, J. Harrison. Bayesian
> Optimization via Continual Variational Last Layer Training
> <https://arxiv.org/abs/2412.09477>. International Conference on Learning
> Representations (ICLR), 2025.
>
> [2] J. Harrison, J. Willes, J. Snoek. Variational Bayesian Last Layers
> <https://arxiv.org/abs/2404.11599>. International Conference on Learning
> Representations (ICLR), 2024.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/pytorch/botorch/pull/2754
> Commit Summary
>
>    - f599471
>    <https://github.com/pytorch/botorch/pull/2754/commits/f599471544eb036b608ff49473f5a08e3633b7e2>
>    add vbll surrogate and notebook
>    - 6b82184
>    <https://github.com/pytorch/botorch/pull/2754/commits/6b82184e84cf3ae79c5e3c9693405ea5a4f3a9c1>
>    update notebook and base implementation
>    - 1b0f899
>    <https://github.com/pytorch/botorch/pull/2754/commits/1b0f8996cb2214e15ccf19ce9d2977fcf64fb5e5>
>    update optim config
>    - 2606ac3
>    <https://github.com/pytorch/botorch/pull/2754/commits/2606ac382334885a990289d395ce840301bda402>
>    add tests for vbll model
>    - 873163f
>    <https://github.com/pytorch/botorch/pull/2754/commits/873163f16f54ef745d533010ce4fbbb59f63ba3a>
>    update tutorial for vblls
>
> File Changes
>
> (6 files <https://github.com/pytorch/botorch/pull/2754/files>)
>
>    - *A* botorch_community/acquisition/bll_thompson_sampling.py
>    <https://github.com/pytorch/botorch/pull/2754/files#diff-e1d4fb223629e80cd1469b9e3965e95cbde8c7cd3aabc2a588e7a859c75fd8d4>
>    (133)
>    - *A* botorch_community/models/vblls.py
>    <https://github.com/pytorch/botorch/pull/2754/files#diff-7c675eabda341230ee44f210363a77bc8ec06e8a1f5d905bcec7a614eef78502>
>    (418)
>    - *A* botorch_community/posteriors/__init__.py
>    <https://github.com/pytorch/botorch/pull/2754/files#diff-c655d31ec372b021b283dd2eb9469ffe897917b48cb753120689904c029241d2>
>    (4)
>    - *A* botorch_community/posteriors/bll_posterior.py
>    <https://github.com/pytorch/botorch/pull/2754/files#diff-a45c3f4d1096e4c276c5269ed330a44650361f816d5e5830eb29d7a0b38adf94>
>    (54)
>    - *A* notebooks_community/vbll_thompson_sampling.ipynb
>    <https://github.com/pytorch/botorch/pull/2754/files#diff-4faf6dfab200cc0e3b077731b3438146ce836eecef404a848be04a27aa5b3020>
>    (803)
>    - *A* test_community/models/test_vblls.py
>    <https://github.com/pytorch/botorch/pull/2754/files#diff-17cdea2ea479a1ee1616e238b6b1a155df7a92b3223d281ce7f37d4beb9cda61>
>    (186)
>
> Patch Links:
>
>    - https://github.com/pytorch/botorch/pull/2754.patch
>    - https://github.com/pytorch/botorch/pull/2754.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/botorch/pull/2754>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAW34NCIDINOF67WZXXLGL2Q45FJAVCNFSM6AAAAABXTNRYUOVHI2DSMVQWIX3LMV43ASLTON2WKOZSHA3DSMRWGIYDIMQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

Balandat commented:
Thanks for putting this up, @brunzema. The notebook looks great, and I plan to review this PR in more detail over the next day or two. 

Regarding the dependency on vbll: Right now it looks like the code only uses ~120 lines of pure torch code from the vbll repo (namely this 
https://github.com/VectorInstitute/vbll/blob/main/vbll/layers/regression.py#L34-L149 plus the minimal function https://github.com/VectorInstitute/vbll/blob/main/vbll/utils/distributions.py#L94-L98). It seems questionable to me to take that dependency, especially since the vllb repo doesn't include unit tests and/or CI. We spend a nontrivial amount of time fixing issues with downstream dependencies, so we're really careful about adding them only if really necessary.

My preference would be to move the relevant pieces of the vbll code mentioned above into a helper module (and clearly attribute the source there, of course) so we can avoid the dependency for now. If we do end up expanding the functionality and use additional features from vbll, then I'd be happy to reconsider (provided the vbll repo adds proper unit tests and a CI setup).

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,133 @@
+from typing import Optional
```
We'll need to add this license header to each of the files in this PR.

```suggestion
#!/usr/bin/env python3
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

from typing import Optional
```

brunzema replied:
added this to all files 👍

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,133 @@
+
+from botorch_community.models.vblls import AbstractBLLModel
+
+torch.set_default_dtype(torch.float64)
```
This will have side effects on other code as well. Can you instead explicitly set the dtype at the places in the code (shouldn't be too many occurences).

brunzema replied:
changed this here as well as the other occurrence 👍

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,133 @@
+    with optional optimization to refine the sampling process.
+
+    Args:
+        model (AbstractBLLModel):
+            The VBLL model from which posterior samples are drawn. Must be an instance of `AbstractBLLModel`.
```
Formatting consistency nit (applies throughout this PR). Also, we generally don't include the type in the docstring since this is evident from the type signature (and will be included automatically in the sphinx API docs).
```suggestion
        model: The VBLL model from which posterior samples are drawn. Must be an instance of `AbstractBLLModel`.
```

brunzema replied:
tried to make the docstring as consistent as possible and removed the dtypes

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,133 @@
+
+    def __init__(
+        self,
+        model,
```
```suggestion
        model: AbstractBLLModel,
```

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,133 @@
+    This class provides functionality to sample from the posterior distribution of a BLL model,
+    with optional optimization to refine the sampling process.
+
+    Args:
```
Consistency nit: We usually add the docstring for the `Args` to the docstring of the `__init__()` method.

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,133 @@
+            self.ub = bounds[1, :].cpu()
+            self.bounds = [tuple(bound) for bound in bounds.T.cpu().tolist()]
+
+    def __call__(self, X_cand: torch.Tensor = None, num_samples: int = 1):
```
```suggestion
    def __call__(self, X_cand: torch.Tensor = None, num_samples: int = 1) -> torch.Tensor:
```

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,133 @@
+        if self.discrete and X_cand is None:
+            raise ValueError("X_cand must be provided if `discrete` is True.")
+
+        sampled_functions = [self.model.sample() for _ in range(num_samples)]
```
Is there any significance to sampling these first and then looping over them? As compared to the alternative of sampling these one-by-one in the loop below?

brunzema replied:
no, there is no difference. I now changed it to the one-by-one for readability :)

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,133 @@
+
+        # get max of sampled functions at candidate points for each function
+        for i, f in enumerate(sampled_functions):
+            # Note that optimization overwrites sampling-based approach
```
I'm not sure I understand this comment, can you make this more clear?

brunzema replied:
hope the new version clarifies this. I also added a `logger.warning` that if `self.discrete=True` but no `X_cand` are provided, then it will default to numerically optimizing the sample path.

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,133 @@
+        for i, f in enumerate(sampled_functions):
+            # Note that optimization overwrites sampling-based approach
+            if not self.discrete:
+                X_cand = torch.empty(
```
Seems like this whole block inside this if statement could be moved into a helper function to make the code more readable.

brunzema replied:
agree, with a helper function its much nicer; changed it accordingly 👍

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,133 @@
+                    )
+
+                    if not res.success:
+                        print(f"Optimization failed with message: {res.message}")
```
should we make this a `logger.warn`?

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,133 @@
+                    X_cand[j, :] = torch.from_numpy(res.x)
+                    Y_cand[j] = torch.tensor([-res.fun])
+
+                # select the best candidate
+                X_next[i, :] = X_cand[Y_cand.argmax()]
+            else:
+                # sampling based approach on candidate points
+                Y_cand = f(X_cand)
+                X_next[i, :] = X_cand[Y_cand.argmax()]
```
```suggestion
            else:
                # sampling based approach on candidate points
                Y_cand = f(X_cand)
                
            # select the best candidate
            X_next[i, :] = X_cand[Y_cand.argmax()]
```

Balandat commented on botorch_community/models/vblls.py:
```diff
@@ -0,0 +1,418 @@
+
+from botorch_community.posteriors.bll_posterior import BLLPosterior
+
+import vbll
```
As per [my earlier comment](https://github.com/pytorch/botorch/pull/2754#issuecomment-2682140981), I would prefer that at this point we copy the small amount of code required from vbll to a helper module.

brunzema replied:
As discussed, I now included a file with the core vbll functionality that is relevant for BO. 👍

Balandat reviewed (commented):
Unit tests are failing b/c vbll is not installed in the CI. This will not be necessary if we move the minimal code required in a helper module that we include here.

Balandat commented on notebooks_community/vbll_thompson_sampling.ipynb:
```diff
@@ -0,0 +1,803 @@
+    "$$\n",
+    "where $\\varepsilon \\sim \\mathcal{N}(0, \\sigma^2)$.\n",
+    "VBLL models use a variational posterior distribution on the weights as $\\mathbf{w} \\sim q(\\mathbf{w})$ with $q(\\mathbf{w}) = \\mathcal{N}(\\bar{\\mathbf{w}}, S)$ where $\\bar{\\mathbf{w}}$ is the mean and $S$ the full covariance of a multivariate normal distribution (for more information, see [1,2]).\n",
+    "Through conjugany, we yield the posterior predictive\n",
```
```suggestion
    "Through conjugacy, we yield the posterior predictive\n",
```

Balandat commented on notebooks_community/vbll_thompson_sampling.ipynb:
```diff
@@ -0,0 +1,803 @@
+    "\n",
+    "## The VBLLs on Toy Data\n",
+    "\n",
+    "Before using the VBLL model in the context of Bayesian Optimization, let's see how the interface of the model looks. The current implementation allows for passing a full backbone to the model (which could be anything). If no backbone is provided, a standard MLP is created. For training, [1] discusses continual learning approaches which will be added to the implementation, but in the followin ...
```
anything? Can you be a bit more specific what API it needs to satisfy?

Balandat commented on notebooks_community/vbll_thompson_sampling.ipynb:
```diff
@@ -0,0 +1,803 @@
+    "# lets print the model, we can see the MLP backbone and the VBLL regression head\n",
+    "print(model)\n",
+    "\n",
+    "# fit the model on the data -- it is also possible to specify the optimizer throught `optimizer_settings` (see docstring)\n",
```
split this line for readability

Balandat commented on notebooks_community/vbll_thompson_sampling.ipynb:
```diff
@@ -0,0 +1,803 @@
+    "For a Thompson sample, we simply sample from the variational posterior of $\\mathbf{w}$ at iteration $t$ and then construct a sample from the predictive $\\hat{f}$ as a generalized linear model:\n",
+    "\n",
+    "\\begin{aligned}\n",
+    "&\\textcircled{1} \\quad \\hat{\\mathbf{w}} \\sim q(\\mathbf{w}) \\\\\n",
+    "&\\textcircled{2} \\quad \\hat{f} (\\mathbf{x}) \\coloneqq \\hat{\\mathbf{w}}^\\top \\phi_{\\theta} (\\mathbf{x}).\n",
```
This doesn't render properly (at least on the GitHub viewer)

Balandat commented on notebooks_community/vbll_thompson_sampling.ipynb:
```diff
@@ -0,0 +1,803 @@
+    "    x_test = torch.linspace(0, 1, 100)\n",
+    "    with torch.no_grad():\n",
+    "        posterior = model.posterior(x_test.view(-1, 1))\n",
+    "    mean, std = posterior.mean.squeeze(), posterior.variance.sqrt().squeeze()\n",
```
Nit: Make this two lines for readability

Balandat commented on notebooks_community/vbll_thompson_sampling.ipynb:
```diff
@@ -0,0 +1,803 @@
+    "test_X = torch.linspace(0, 1, 100)\n",
+    "with torch.no_grad():\n",
+    "    posterior = model.posterior(test_X.view(-1, 1))\n",
+    "mean, std = posterior.mean.squeeze(), posterior.variance.sqrt().squeeze()\n",
```
Nit: Make this two lines for readability

Balandat commented on notebooks_community/vbll_thompson_sampling.ipynb:
```diff
@@ -0,0 +1,803 @@
+    "    x,\n",
+    "    y,\n",
+    "    new_data=None,\n",
+    "    show_opjective=False,\n",
```
```suggestion
    "    show_objective=False,\n",
```

brunzema replied:
...again, ups and thank you!

Balandat commented on notebooks_community/vbll_thompson_sampling.ipynb:
```diff
@@ -0,0 +1,803 @@
+    "\n",
+    "    ax.plot(x_test, mean, label=\"Posterior predictive\", color=\"tab:blue\")\n",
+    "\n",
+    "    # Posterior samples\n",
```
The TS samples shown are not the samples that were optimized to obtain the next candidate point. This initially caused some confusion; can you make the plots in a way so that this is consistent (i.e. that we show the set of TSs that were optimized to obtain the next candidate)?

Balandat commented on notebooks_community/vbll_thompson_sampling.ipynb:
```diff
@@ -0,0 +1,803 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhAAAAF2CAYAAAA/RaFTAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAzSNJREFUeJzsnQd8VGXWxs/0mt5ISKH3pogKKIiIiL0rllXX/tl27e661lVX195dey8odkRBARFQuvROgPSeyfR2v99z7txh0mdCGuH9746TKblz52aY99xznvMclSRJEgkEAoFAIBDEgDqWJwsEAoFAIBAAEUAIBAKBQCCIGRFACAQCgUAgiBkRQAgEAoFAIIgZEUAIBAKBQCCIGRF ...
```
It would be really nice to run both TS and logEI and then show these in a two-panel figure to have the comparison. This would also ensure that both code paths work properly

brunzema replied:
I changed it to loop over both logEI and TS and then show the models for all iterations. I think the model "evolution" is very similar. I wanted to keep the main BO loop as simple as possible rather than optimizing in parallel with both acquisition function for easier reuse of the loop. If you think two-panel figures would be more interesting, I'm happy to make that change!

Balandat commented on notebooks_community/vbll_thompson_sampling.ipynb:
```diff
@@ -0,0 +1,803 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhAAAAF2CAYAAAA/RaFTAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAzSNJREFUeJzsnQd8VGXWxs/0mt5ISKH3pogKKIiIiL0rllXX/tl27e661lVX195dey8odkRBARFQuvROgPSeyfR2v99z7txh0mdCGuH9746TKblz52aY99xznvMclSRJEgkEAoFAIBDEgDqWJwsEAoFAIBAAEUAIBAKBQCCIGRFACAQCgUAgiBkRQAgEAoFAIIgZEUAIBAKBQCCIGRF ...
```
Also here, comparing TS with logEI would be interesting (though not really meaningful if we only do one sample, but we can point out that caveat prominently.

Balandat replied:
This seems to be producing a numpy `DeprecationWarning`, can  you update this to avoid this?

Balandat commented on notebooks_community/vbll_thompson_sampling.ipynb:
```diff
@@ -0,0 +1,803 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhAAAAF2CAYAAAA/RaFTAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAzSNJREFUeJzsnQd8VGXWxs/0mt5ISKH3pogKKIiIiL0rllXX/tl27e661lVX195dey8odkRBARFQuvROgPSeyfR2v99z7txh0mdCGuH9746TKblz52aY99xznvMclSRJEgkEAoFAIBDEgDqWJwsEAoFAIBAAEUAIBAKBQCCIGRFACAQCgUAgiBkRQAgEAoFAIIgZEUAIBAKBQCCIGRF ...
```
I don't know how much extra value this adds over the previous demonstration, esp. if we don't use it to compare performance. I suggest removing the Ackley part to focus on the key messages / examples.

brunzema replied:
sounds good, removed Ackley

Balandat reviewed (commented):
oops, submitted prematurely. Here is the rest of the review.

brunzema commented on botorch_community/models/vblls.py:
```diff
@@ -0,0 +1,418 @@
+        self.backbone = backbone
+        self.sampled_params = sampled_params
+
+    def forward(self, x: Tensor) -> Tensor:
```
Added the dimensions for the forward method here and also for the vbll networks 👍

brunzema commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,133 @@
+        self,
+        model,
+        num_restarts: int = 10,
+        bounds: Optional[torch.Tensor] = None,
```
Adapted the style in the updated PR

brunzema commented on botorch_community/models/vblls.py:
```diff
@@ -0,0 +1,418 @@
+        )
+
+        # Make dataloader based on train_X, train_y
+        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
Changed it to pulling the device from `self.model` which can explicitly be initialized with a device

brunzema commented on botorch_community/models/vblls.py:
```diff
@@ -0,0 +1,418 @@
+        # load best model
+        if best_model_state is not None:
+            self.model.load_state_dict(best_model_state)
+            print("Early stopping at epoch ", epoch, " with loss ", best_loss)
```
nice, yes moved all print statements to the logger

brunzema commented on botorch_community/models/vblls.py:
```diff
@@ -0,0 +1,418 @@
+        K = self.num_outputs
+        mean = mean.reshape(B, N * K)
+
+        # Cov must be `(B, Q*K, Q*K)`
```
Ups, yes!

brunzema commented on botorch_community/posteriors/bll_posterior.py:
```diff
@@ -0,0 +1,54 @@
+
+
+class BLLPosterior(Posterior):
+    def __init__(self, posterior, model, X, output_dim):
```
Included types here, but I had to do a construction with `TYPE_CHECKING` from  the `typing` package to avoid circular imports

brunzema commented on botorch_community/models/vblls.py:
```diff
@@ -0,0 +1,418 @@
+        return BLLPosterior(post_pred, self, X, self.num_outputs)
+
+    @abstractmethod
+    def sample(self, sample_shape: Optional[torch.Size] = None):
```
added `nn.Module` as return type

brunzema commented on test_community/models/test_vblls.py:
```diff
@@ -0,0 +1,186 @@
+
+            X, y = _reg_data_singletask(d)
+            optim_settings = {
+                "num_epochs": 10,
```
no, not necessary--changed it to 1 and increased the learning rate to make sure we get significant changes

brunzema commented:
@Balandat Thank you again for the detailed review--I really appreciate it! I’ve updated the PR to address all your points, but please let me know if any further changes are needed. I’m happy to make any additional updates!

Biggest change is of course the added code from the vbll package, let me know if the way I know included it is ok (additional file + credit at the top).

Balandat commented:
@brunzema thanks a lot for the updates - will review this within the next couple of days!

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,158 @@
+        return -grad_f(torch.from_numpy(x).to(device)).detach().cpu().numpy()
+
+    for j in range(num_restarts):
+        # generate initial condition from within the bounds
+        x0 = np.random.rand(f.num_inputs)
+        x0 = lb + (ub - lb) * x0
```
Nit: Could do this once in a vectorized fashion outside the loop and then just index in the loop (more a µ-optimization though)

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,158 @@
+            func, x0, jac=grad_func, bounds=bounds, method="L-BFGS-B"
+        )
+
+        if not res.success:
```
Does this work reliably? I've run into cases where the line search fails in odd ways sometimes, if there is a chance of this happening in any of the samples and we use a lot of samples then this could cause frequent failures. If this doesn't happen then great, o/w we may want to handle this, possibly by retrying the optimization

Balandat replied:
By this pointer you mean my comment above? I guess right now you're not erroring out but just warning, so that should be fine. Though if the majority of of the `num_restarts` have an optimization failure that's probably not good.

brunzema replied:
I've now added an error raise in case none of the initializations result in a successful optimization--just to be sure.

Balandat commented on botorch_community/models/vbll_helper.py:
```diff
@@ -0,0 +1,487 @@
+        return DenseNormalPrec(self.loc.squeeze(idx), self.tril.squeeze(idx))
+
+
+cov_param_dict = {
```
Can you move this up to the top of the module and capitalize this to indicate this is a constant?

Balandat commented on botorch_community/models/vbll_helper.py:
```diff
@@ -0,0 +1,487 @@
+
+
+# Credit to https://github.com/brentyi/fannypack/blob/2888aa5d969824ac1e1a528264674ece3f4703f9/fannypack/utils/_math.py
+def cholesky_inverse(u: torch.Tensor, upper: bool = False) -> torch.Tensor:
```
`torch.cholesky_inverse()` supports batch dimensions since early 2021 (https://github.com/pytorch/pytorch/pull/50269), so I don't think function is needed here

brunzema replied:
true, thank you!

Balandat commented on botorch_community/models/vbll_helper.py:
```diff
@@ -0,0 +1,487 @@
+        return Normal(self.loc.squeeze(idx), self.scale.squeeze(idx))
+
+
+class DenseNormal(torch.distributions.MultivariateNormal):
```
Side note: the term "dense" normal for MVN is a bit confusing here, but I guess that's fine

brunzema replied:
Yeah, I wanted to keep consistency with the vbll package to avoid confusion. But I see how avoiding confusion can lead to more confusion… lol.

Balandat replied:
Maybe add a comment to the docstring clarifying this?

Balandat commented on botorch_community/models/vbll_helper.py:
```diff
@@ -0,0 +1,487 @@
+        warnings.warn(
+            "Direct matrix inverse for dense covariances is O(N^3), consider using eg inverse weighted inner product"
+        )
+        return tp(torch.linalg.inv(self.scale_tril)) @ torch.linalg.inv(self.scale_tril)
```
Ugh, this is pretty bad. You're calling `inv` on the same tensor twice. Also, you already have the Cholesky decomposition here, why are you calling `inv` in the first place? You can do this with a simple triangular solve.
```suggestion
        Eye = torch.eye(self.scale_tril.shape[-1], device=self.scale_tril.device, dtype=self.scale_tril.dtype)
        W = torch.linalg.solve_triangular(self.scale_tril, Eye, upper=False)
        return tp(W) @ W
```

brunzema replied:
uff, thank you! Will also push this to the vbll repo

Balandat commented on botorch_community/models/vbll_helper.py:
```diff
@@ -0,0 +1,487 @@
+
+    def precision_weighted_inner_prod(self, b, reduce_dim=True):
+        assert b.shape[-1] == 1
+        prod = (torch.linalg.solve(self.scale_tril, b) ** 2).sum(-2)
```
should use `solve_triangular` here as well

Balandat commented on botorch_community/models/vbll_helper.py:
```diff
@@ -0,0 +1,487 @@
+    def chol_covariance(self):
+        raise NotImplementedError()
+
+    @property
+    def covariance(self):
+        return self.cov_factor @ tp(self.cov_factor) + torch.diag_embed(self.cov_diag)
```
seems like the `torch.distributions.LowRankMultivariateNormal` does this just fine, why overwrite?

brunzema replied:
Actually not super sure if there was a deeper reason here, I will double check!

Balandat replied:
any intel on this?

brunzema replied:
yeah, as you said it does exactly the same so I removed it. This was probably from earlier iterations of the vbll repo.

Balandat commented on botorch_community/models/vbll_helper.py:
```diff
@@ -0,0 +1,487 @@
+                - 0.5 * self.wishart_scale * noise.trace_precision
+            )
+            total_elbo = torch.mean(pred_likelihood - trace_term)
+            total_elbo += self.regularization_weight * (wishart_term - kl_term)
```
in-pace updates can be problematic for backprop, suggest avoiding that
```suggestion
            total_elbo = total_elbo + self.regularization_weight * (wishart_term - kl_term)
```

Balandat commented on botorch_community/posteriors/bll_posterior.py:
```diff
@@ -0,0 +1,69 @@
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from botorch_community.models.vblls import AbstractBLLModel
```
is this still needed if you use `from __future__ import annotations` in all the modules to do postponed type evaluations? We should generally use that.

Balandat reviewed (changes requested):
Thanks for the updates and in particular for avoiding another external dependency! I left a number of inline comments, here are some higher level ones: 
1. Please rebase this on a recent version of `master` (the base of commit of this PR is pretty old)
2. Please address flake8 (incl. line length) and import sorting errors - you can install the `pre-commit` hooks to make sure that your code confirms to the standard (see https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pre-commit-hooks)
3. I spotted some some rather problematic numerical code in the vbll helpers, let's update that (highlighted inline), shall we?

brunzema commented:
@Balandat no, the delay is fully on me, sorry! Everything is pretty much done, flake8 is adressed and also with the import sorting +  `from __future__ import annotations` I no longer require the not so nice type_checking condition 👍 I will make a fresh rebase tomorrow and a final pass and hopefully in the morning your time, I have submitted the updated PR :)

Balandat commented:
Excellent!

brunzema commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,158 @@
+            raise ValueError("X_cand must be provided if `discrete` is True.")
+
+        if X_cand is not None and not self.discrete:
+            logger.warning(
```
yeah makes sense, changed it to raise an error

brunzema commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,158 @@
+            f = self.model.sample()
+
+            # NOTE: If self.discrete is False, we will always numerically optimize the sample
+            # path for X_next. X_cand is ignored in this case.
```
Happy to include this down the line! Seems very relevant for incorporating some one-hot encoding-like inputs + some parameters in R^d. I am mainly thinking about some autoML applications. Would it be ok to include this in a later PR?

brunzema commented:
@Balandat I updated the PR. Not sure why I thought the importing was fixed yesterday—I still had the type checking in place. I noticed that some files in the main repo also use it, so it should be fine?

That said, I think a cleaner approach (which I’ve implemented in the updated PR) is to extract the abstract Bayesian last-layer (BLL) class and use it as a parent class/interface. This should also make it more extensible for future BLL models.

```
botorch_community
|-- acquisition
|   |-- bll_thompson_sampling.py # AbstractBLL to specify "interface"
|-- models
|   |-- blls.py # Define AbstractBLL model
|   |-- vblls.py # VBLL inherit from AbstractBLL
|-- posteriors
|   |-- bll_posterior.py # AbstractBLL to specify "interface"
...
```

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,173 @@
+from torch.func import grad
+
+
+class BLLMaxPosteriorSampling:
```
There is a [`SamplingStrategy`](https://github.com/pytorch/botorch/blob/main/botorch/generation/sampling.py#L39) interface here that that defines the API for this (and subclasses `torch.nn.Module`) but doesn't support not providing an `X_cand`. I assume this is why you didn't subclass from that? 

I don't think it's necessary to use this if it would require changes to `SamplingStrategy` but maybe leave a comment here that we may want to consider doing that in the future.

Balandat commented on botorch_community/acquisition/bll_thompson_sampling.py:
```diff
@@ -0,0 +1,173 @@
+
+    def __call__(
+        self, X_cand: torch.Tensor = None, num_samples: int = 1
+    ) -> torch.Tensor:
```
Can you add a docstring describing this method, including the expected input and return shapes of the tensors?

Balandat reviewed (commented):
Thanks! Overall this looks great. 

Before I merge this in, could you please increase unit test coverage - looks like quite a lot of things are [not covered](https://github.com/pytorch/botorch/pull/2754/checks?check_run_id=39657261673) by tests, including some important parts such as `BLLPosterior.rsample()`.

The tutorial failure is unrelated, we'll fix this on our end.

brunzema commented:
hey @Balandat , again took a while--sorry about that! I tried to push the test coverage to 100%. Only places that I am unsure about are the following:

1) After how small discussion, I now added an error to the numerical optimization of the posterior samples paths (See https://github.com/pytorch/botorch/pull/2754#discussion_r2020170659):

```python
def _optimize_sample_path():
    ...
    
    optimization_successful = False
    for j in range(num_restarts):
        # map to bounds
        x0 = lb + (ub - lb) * x0s[j]

        # optimize sample path
        res = scipy.optimize.minimize(
            func, x0, jac=grad_func, bounds=bounds, method="L-BFGS-B"
        )

        # check if optimization was successful
        if res.success:
            optimization_successful = True
        if not res.success:
            logger.warning(f"Optimization failed with message: {res.message}")

        # store the candidate
        X_cand[j, :] = torch.from_numpy(res.x).to(dtype=torch.float64)
        Y_cand[j] = torch.tensor([-res.fun], dtype=torch.float64)

    if not optimization_successful:
        raise RuntimeError("All optimization attempts on the sample path failed.")
        
 ```
Here I am unsure how to actually test this as this never happened to be before. I am inclined to put a `# pragma: no cover` here. What do you think?

2) Should I write a separate test for the vbll helpers? The not covered lines are essentially all properties and the important parts (different parameterization of the VBLL head) are now all covered in the VBLL test. What is your opinion? Best place would be in `utils`?

Balandat commented:
Awesome, thanks for pushing this along. Once we cover the last few lines with tests this can go in. 

> Here I am unsure how to actually test this as this never happened to be before. I am inclined to put a # pragma: no cover here. What do you think?

I'd recommend having a lightweight mocked test that mocks out `scipy.optimize.minimize` to return a mocked `OptimizeResult` where `success = False`.

> Should I write a separate test for the vbll helpers? The not covered lines are essentially all properties and the important parts (different parameterization of the VBLL head) are now all covered in the VBLL test. What is your opinion? Best place would be in utils?

Yes, having those tests separately would be great. Looks like the misses are mainly in `botorch_community/models/vbll_helper.py`, so I would just test these in a new `test_community/models/test_vbll_helper.py` module.

Balandat commented:
Thanks @brunzema for the awesome contribution!

eytan commented:
Thanks Paul for the thorough contribution!  Congrats on landing the paper at ICLR and glad to see we could get it into BoTorch before the conference :)

brunzema commented:
Thank you guys also for the effort and the thorough reviews—I really appreciate it! Now let’s see what people build with it. 👍  Looking forward to seeing you (and/or others from the team) at ICLR!

Balandat commented:
> Now let’s see what people build with it. 👍

Indeed, looking forward to it! 

> Looking forward to seeing you (and/or others from the team) at ICLR!

I don't think anyone from the team is attending ICLR this time around - but hope to see you at a future conference!
